lib/ansible/plugins/lookup/openshift.py

```diff
@@ -0,0 +1,248 @@
+        required: False
+      pretty:
+        description:
+          - Whether or not to prettify the output.  This is useful for
+          - debugging.
+        required: False
+        default: False
+      labelSelector:
+        description:
+          - Additional labels to include in the query.
+        required: False
+      fieldSelector:
+        description:
+          - Specific fields on which to query.
+        required: False
+      resourceVersion:
+        description:
+          - Query for a specific resource version.
+        required: False
+"""
+
+EXAMPLES = """
+- name: Get Project {{ project_name }}
+  set_fact:
+    project_fact:  "{{ lookup('openshift',
+                       'Project',
+                       host=inventory_host,
+                       token=hostvars[inventory_host]['ansible_sa_token'],
+                       resource_name=project_name,
+                       validate_certs=validate_certs) }}"
+- name: Get All Service Accounts in a Project
+  set_fact:
+    service_fact: "{{ lookup('openshift',
+                      'ServiceAccount',
+                      host=inventory_host,
+                      token=hostvars[inventory_host]['ansible_sa_token'],
+                      namespace=project_name,
+                      validate_certs=validate_certs) }}"
+"""
+
+RETURN = """
+  _list:
+    description:
+      - An object definition or list of objects definitions returned from OpenShift.
+    type: dict
+"""
+
+import json
+from ansible.errors import AnsibleError
+from ansible.plugins.lookup import LookupBase
+from ansible.module_utils import urls
+from ansible.module_utils.six.moves import urllib
+from ansible.module_utils.six.moves import urllib_error
+from ansible.module_utils.six.moves.urllib.parse import urlencode
+
+
+class OcpQuery(object):
+    def __init__(self, host, port, token, validate_certs):
+        self.apis = ['api', 'oapi']
+        self.token = token
+        self.validate_certs = validate_certs
+        self.host = host
+        self.port = port
+        self.kinds = {}
+        bearer = "Bearer " + self.token
+        self.headers = {"Authorization": bearer}
+        self.build_facts()
+
+    def build_facts(self):
+
+        for api in self.apis:
+            url = "https://{0}:{1}/{2}/v1".format(self.host, self.port, api)
+            try:
+                response = urls.open_url(url=url,
+                                         headers=self.headers,
+                                         validate_certs=self.validate_certs,
+                                         method='get')
+            except urllib_error.HTTPError as e:
+                try:
+                    body = e.read()
+                except AttributeError:
+                    body = ''
+                raise AnsibleError("OC Query raised exception with code {0} and message {1} against url {2}".format(e.code, body, url))
+
+            for resource in json.loads(response.read())['resources']:
```

Comment:
you should use to_text() (also from ansible.module_utils._text) around response.read() here.  Pretty sure that returns a byte string on both python2 and python3.

Reply:
Since this is setting data, you also want to use an error handler like this:
``` python
from ansible.module_utils._text import to_text
[...]
for resource in json.loads(to_text(response.read(), errors='surrogate_or_strict'))['resources']:
```

The default error handler is replace which is good for things displayed to the user but not good for preserving data.